[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1bv9Di1PmgvNHiL8Ue2LWTjekvfYXEQ8k?usp=sharing)

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

14 88
102

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

1337 1488
2825


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


##### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

10
2 3 5 7 11 13 17 19 23 29
667


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {0..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
10
200000 23 200000 0 45 1 4 6 123 32131

Writing input4.txt


In [ ]:
200000 * 200000

40000000000

In [ ]:
%%writefile output4.txt
40000000000

Writing output4.txt


In [ ]:
%%writefile input5.txt
5
3 1 200000 45 20000

Writing input5.txt


In [ ]:
200000 * 20000

In [ ]:
%%writefile output5.txt
4000000000

Writing output5.txt


In [ ]:
%%writefile input6.txt
15
199999 234 0 234 3000 124511 400000 4523 7845 121314 89677 453123 523213 7453 0

Writing input6.txt


In [ ]:
453123 * 523213

237079844199

In [ ]:
%%writefile output6.txt
237079844199

Writing output6.txt


In [ ]:
%%writefile input7.txt
2
900000 5

Writing input7.txt


In [ ]:
900000 * 5

4500000

In [ ]:
%%writefile output7.txt
4500000

Writing output7.txt


In [ ]:
%%writefile input8.txt
19
28543 11213 7663 35730 28443 14637 19039 25441 20156 16505 30649 23923 25435 18913 29108 25409 17174 143 11733

Writing input8.txt


In [ ]:
35730 * 30649

1095088770

In [ ]:
%%writefile output8.txt
1095088770

Writing output8.txt


In [ ]:
%%writefile input9.txt
50
59918 10147 58122 15353 69460 86114 18171 45377 79182 68269
33235 50458 37455 14672 78008 28725 41017 56580 49624 29429
23106 12679 29952 26702 84626 25185 45291 89627 38901 23801
53468 80883 68361 72442 87880 34449 17123 56210 52144 46789
62951 12690 46326 85744 11059 36021 49412 11731 21735 33212

Writing input9.txt


In [ ]:
89627 * 87880

7876420760

In [ ]:
%%writefile output9.txt
7876420760

Writing output9.txt


In [ ]:
%%writefile input10.txt
60
16695 50605 42530 57855 24236 24347 36570 58107 25949 26939
10769 54063 66323 23606 22007 37126 51412 14381 43186 12283
34525 55947 52947 66780 14485 41951 63416 32926 27342 64352
37077 27768 27124 54245 63911 28044 12350 64340 57079 51652
38353 17289 45777 45391 22221 60764 26161 36591 12275 49150
123888 129855 116245 96360 107581 124692 133711 82847 116602 103602

Writing input10.txt


In [ ]:
133711 * 129855

17363041905

In [ ]:
%%writefile output10.txt
17363041905

Writing output10.txt


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {0..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing stress_test.sh


In [ ]:
#Running tests cases. Check "/files" for "outputN.txt" where N is the Nth output for the Nth "inputN.txt" (e.g. input4.txt/output4.txt)
!chmod 744 stress_test.sh
!./stress_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	5m1.546s
user	4m58.824s
sys	0m0.099s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m0.912s
user	4m57.793s
sys	0m0.135s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound

Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 ////INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);
    execints++;

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right) \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

571


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$

In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

601


For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt
n= 10

572


For the input11.txt
```
200000
1 2 ..  199999 200000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177092


In [ ]:
# Estimated time by instruction
295/200001000021


1.4749926248820013e-09

In [ ]:
# Estimated number of instructions by second
1 / (295/200001000021)

677969491.5966102

# Estimate the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.864s
user	0m0.858s
sys	0m0.002s


## Python

In [ ]:
%%writefile max_pairwise_product_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(n):
            if i != j:
                aux_product = l[i] * l[j]
                if aux_product > max_product:
                    max_product = aux_product
    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Writing max_pairwise_product_py.py


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("11000\n")
        for i in range(1, 11001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
# Verify #Not surprising Python version is slower...
%%time
!python max_pairwise_product_py.py < input14.txt

120989000
CPU times: user 150 ms, sys: 27 ms, total: 177 ms
Wall time: 23.5 s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound

Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input11.txt

39999800000

real	2m21.327s
user	2m19.779s
sys	0m0.068s


In [ ]:
%%writefile max_pairwise_product_n2_py.py
#Python3
def max_pairwise_product(l):
    max_product = 0
    aux_product = 0
    n = len(l)
    for i in range(n):
        for j in range(i + 1, n):
            aux_product = l[i] * l[j]
            if max_product < aux_product:
                max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product(l))

Overwriting max_pairwise_product_n2_py.py


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T (n) = 11 + 7n + \left[2(n-1)n+1, 2.5(n-1)n\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```

$$T(10) \in [262, 306] $$

In [ ]:
n = 10
print(11 + 7*(10) + 2*(10)*(9) + 1, 11 + 7*(10) + 2.5*(10)*(9))

262 306.0


For the input11.txt
```
200000
1 2 . . . 199999 200000
```

$$T(200000) \in [80001000012, 100000900011] $$

In [ ]:
print(11 + 7*(200000) + 2*(200000)*(199999) + 1, 11 + 7*(200000) + 2.5*(200000)*(199999))

80001000012 100000900011.0


In [ ]:
#ins
estimated_time = 195 / 100000900009
estimated_time

1.9499824499824516e-09

In [ ]:
#ins per sec (aprox.)
ins_per_sec =  1 / estimated_time
ins_per_sec

512825128.2512821

###Correctness tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long.cpp -o max_pairwise_product_long

for i in {0..10}
do
    ./max_pairwise_product_long < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_n2_py.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

###Estimated maximum that can be solved in 1 second
$$(11 + 7n + 2.5(n-1)n)*1.47*10^{-9} = 1s$$
$$ n \approx 16400 $$

In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 16400 <<"\n";  // n
 for (int i=1; i <16401; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

268943600

real	0m0.894s
user	0m0.880s
sys	0m0.005s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("16400\n")
        for i in range(1, 16401):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
#Slower. Not a surprise.
%%time
!python max_pairwise_product_n2_py.py < input14.txt

268943600
CPU times: user 140 ms, sys: 11.9 ms, total: 152 ms
Wall time: 19.5 s


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> v(n);                                   //n
    for(int i = 0; i < n; i++){                              //1 + 2n
        std::cin >> v[i];                                    //n
    }

    std::sort(v.begin(), v.end());                          //nlogn
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';   //2
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [ ]:
%%writefile max_pairwise_product_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())                                        #1
    l = list(map(int, input().split()))                     #5n
    l.sort()                                                #nlogn at least
    print(l[-1] * l[-2])                              #2

Overwriting max_pairwise_product_sort.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!./max_pairwise_product_sort < input1.txt   #Original array in input1.txt increasing order [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
!./max_pairwise_product_sort < input2.txt   #Original array in input2.txt decreasing order [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

90
90


In [ ]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [ ]:
%%time
#Given in secs. *10^-3 ms
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 4.75 ms, sys: 808 µs, total: 5.56 ms
Wall time: 106 ms


20 stress tests

In [ ]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 49 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::sort(v.begin(), v.end());
        std::cout << "n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [ ]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

n: 17 Product: 850976
n: 25 Product: 857451
n: 43 Product: 980036
n: 10 Product: 692094
n: 30 Product: 839816
n: 48 Product: 956363
n: 9 Product: 475580
n: 46 Product: 954088
n: 44 Product: 902850
n: 17 Product: 981034
n: 40 Product: 928812
n: 7 Product: 267968
n: 42 Product: 991020
n: 27 Product: 908160
n: 23 Product: 990016
n: 2 Product: 141372
n: 4 Product: 418500
n: 11 Product: 557996
n: 42 Product: 963270
n: 17 Product: 690616


In [ ]:
%%writefile stress_tests_sort_py.py
#Python3
import random
if __name__ == "__main__":
    for i in range(10):
        n = random.randint(4, 186)
        l = [ random.randint(27, 56734) for i in range(n) ]
        l.sort()
        print(f"n: {n} Product: {l[-1] * l[-2]}")

Writing stress_tests_sort_py.py


In [ ]:
!python stress_tests_sort_py.py

n: 128 Product: 3050714716
n: 77 Product: 3117904825
n: 82 Product: 3082037882
n: 114 Product: 3144293307
n: 142 Product: 3148816518
n: 58 Product: 3025599480
n: 66 Product: 3150576371
n: 168 Product: 3117074317
n: 76 Product: 3203502588
n: 17 Product: 2986001436


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                                                   //1
    ins++;
    std::cin >> n;                                           //1
    ins++;
    std::vector<uint64_t> v(n);                                   //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){                              //1 + 2n
        ins += 2;
        std::cin >> v[i];                                    //n
        ins++;
    }

    std::sort(v.begin(), v.end());                           //nlogn at least (nlogn time complexity)
    ins += n * log2(n);
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';    //3
    ins += 3;
    std::cout << "Executed instructions: " << ins << '\n';   // 7 + 4n + ins(nlogn sort). Where ins(nlogn sort) are, at least, nlogn.
    return 0;
}

Overwriting max_pairwise_product_sort_ins.cpp


In [ ]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 79
90
Executed instructions: 79
40000000000
Executed instructions: 45


In [ ]:
!time ./max_pairwise_product_sort_ins < input11.txt

39999800000
Executed instructions: 4321934

real	0m0.126s
user	0m0.115s
sys	0m0.006s


In [ ]:
%%writefile max_pairwise_product_sort_ins.py
#Python3
import math

if __name__ == "__main__":
    ins = 0
    n = int(input())                       #1
    ins += 1
    l = list(map(int, input().split()))    #5n (n input read, n mappings to int, n iterable to list, n comparisons, n splits through input)
    ins += 5 * n
    l.sort()                               #nlogn at least
    ins += n * math.log2(n)
    print(l[-1] * l[-2])             #2
    ins += 2
    print(f"Executed instructions: {int(ins)}\n")                             #3 + 5n + ins(nlogn timsort) where ins(nlogn sort) are, at least, nlogn.

Overwriting max_pairwise_product_sort_ins.py


In [ ]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Executed instructions: 86

90
Executed instructions: 86

40000000000
Executed instructions: 48



In [ ]:
%%time
!python max_pairwise_product_sort_ins.py < input11.txt

39999800000
Executed instructions: 4521931

CPU times: user 6.39 ms, sys: 0 ns, total: 6.39 ms
Wall time: 110 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
#following result of input3.txt. #lower bound. Needs to know the exact number of instructions in each sorting algorithm. Just counts the ones in the nlogn step.
#Estimaded time by instruction and number of instructions
estimated_time = 0.009 / 4521931
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

1.9903001615902584e-09
502436777.77777785


###Estimated maximum that can be solved in 1 second
$$(5 + 4n + 5nlog(n))*1.47*10^{-9} = 1s \ \ \ given \ that \ nlogn \ at \ least$$
$$ n \approx 8'000.000 $$

In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 8000000 <<"\n";  // n
 for (int i=1; i <8000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest

In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input13.txt

268943600

real	0m0.011s
user	0m0.010s
sys	0m0.001s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input16.txt", "w") as f:
        f.write("8000000\n")
        for i in range(1, 8000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
#Finally..
%%time
!python max_pairwise_product_sort.py < input13.txt

268943600
CPU times: user 3.09 ms, sys: 2.01 ms, total: 5.1 ms
Wall time: 105 ms


###Correctness Tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_sort.cpp -o max_pairwise_product_long

for i in {0..10}
do
    ./max_pairwise_product_sort < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_sort.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    long long p_max = 0;                     //1
    long long s_max = 0;                     //1
    for(int i = 0; i < n; ++i) {             //1 + 2n
        if (v[i] > s_max){                    //n
            s_max = v[i];                    //[0, n]
        }
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            p_max = v[i];                    //[0, n]
        }
    }
    return (long long) p_max * s_max;       //2 -> 9 + 4n + 3[0, n]           //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    std::cin >> n;                          //1
    std::vector<int> v(n);                  //n
    for(int i = 0; i < n; i++){             //1 + 2n
        std::cin >> v[i];                   //n
    }
    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n
    return 0;
}

Writing max_pairwise_product_ef.cpp


In [ ]:
%%writefile max_pairwise_product_ef.py
#Python3
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product_f(l))

Writing max_pairwise_product_ef.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [ ]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.003s
sys	0m0.000s


In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input11.txt

39999800000

real	0m0.067s
user	0m0.063s
sys	0m0.003s


In [ ]:
%%time
!python max_pairwise_product_ef.py < input3.txt

40000000000
CPU times: user 5.55 ms, sys: 0 ns, total: 5.55 ms
Wall time: 111 ms


In [ ]:
%%time
!python max_pairwise_product_ef.py < input11.txt

39999800000
CPU times: user 5.81 ms, sys: 45 µs, total: 5.85 ms
Wall time: 204 ms


20 stress tests

In [ ]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();
    ins++;
    long long p_max = 0;
    ins++;
    long long s_max = 0;
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {
        ins += 2;
        ins++;
        if (v[i] > s_max){
            s_max = v[i];
            ins++;
        }
        ins++;
        if (v[i] > p_max){
            s_max = p_max;
            ins++;
            p_max = v[i];
            ins++;
        }
    }

    return (long long) p_max * s_max;
}

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 69 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::cout << "n: " << n << " Product: " << max_pairwise_product_f(v) << '\n';
    }
    return 0;
}

Writing stress_tests_ef_cpp.cpp


In [ ]:
%%writefile stress_tests_ef_py.py
#Python3
import random
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        l = [random.randint(17, 24531) for i in range(n)]
        print(f"n: {n} Product: {max_pairwise_product_f(l)}")

Writing stress_tests_ef_py.py


In [ ]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 36 Product: 863968
n: 57 Product: 980036
n: 60 Product: 898343
n: 31 Product: 927682
n: 13 Product: 762600
n: 51 Product: 954088
n: 12 Product: 683388
n: 26 Product: 981034
n: 70 Product: 960364
n: 14 Product: 858552
n: 61 Product: 993010
n: 52 Product: 963270
n: 10 Product: 631764
n: 32 Product: 994008
n: 21 Product: 960319
n: 24 Product: 828625
n: 2 Product: 74562
n: 54 Product: 904397
n: 8 Product: 886974
n: 23 Product: 911064


In [ ]:
!python stress_tests_ef_py.py

n: 90 Product: 585448560
n: 13 Product: 552092880
n: 71 Product: 573434442
n: 6 Product: 391043435
n: 58 Product: 590690898
n: 48 Product: 546350748
n: 14 Product: 566225610
n: 29 Product: 540369779
n: 77 Product: 591778140
n: 20 Product: 530476040
n: 37 Product: 524940160
n: 45 Product: 536931315
n: 8 Product: 417076998
n: 6 Product: 475998432
n: 52 Product: 582361152
n: 26 Product: 489188698
n: 93 Product: 599269031
n: 20 Product: 465024597
n: 42 Product: 571016712
n: 55 Product: 600028590


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    ins++;
    long long p_max = 0;                     //1
    ins++;
    long long s_max = 0;                     //1
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {             //1 + 2n
        ins += 2;
        ins++;
        if (v[i] > s_max){                    //n
            s_max = v[i];                     //[0, n]
            ins++;
        }
        ins++;
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            ins++;
            p_max = v[i];                    //[0, n]
            ins++;
        }
    }
    ins += 2;
    return (long long) p_max * s_max;       //2 -> 6 + 4n + 3[0, n]          //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    ins++;
    std::cin >> n;                          //1
    ins++;
    std::vector<int> v(n);                  //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){             //1 + 2n
        ins += 2;
        std::cin >> v[i];                   //n
        ins++;
    }

    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n + 6 + 4n + 3[0, n] -> 10 +8n + 3[0, n] -> 10 + 11n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_ef_ins.cpp


In [ ]:
%%writefile max_pairwise_product_ef_ins_py.py
#Python3
ins = 0
def max_pairwise_product_f(l):
    global ins
    p_max = 0                               #1
    ins += 1
    s_max = 0                               #1
    ins += 1
    ins += 1
    for i in l:                             #1 + 2n
        ins += 2
        if i > s_max:                       #n
            s_max = i                       #[0, n]           #[0, n] means that, in the worst case, executes n times
            ins += 1
        if i > p_max:                       #n
            s_max = p_max                   #[0, n]
            ins += 1
            p_max = i                       #[0, n]
            ins += 1

    ins += 2
    return p_max * s_max                    #2       #5 + 4n + 3[0, n] -> 5 + 7n worst

if __name__ == "__main__":
        n = int(input())                                 #1
        ins += 1
        l = list(map(int, input().split()))              #5n
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))                 #1
        print(f"Executed instructions: {ins}")           #-> 2 + 5n + 5 + 7n -> 7 + 12n

Writing max_pairwise_product_ef_ins_py.py


In [ ]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 119
90
Executed instructions: 93
40000000000
Executed instructions: 68


In [ ]:
!time ./max_pairwise_product_ef_ins < input11.txt

39999800000
Executed instructions: 2200009

real	0m0.077s
user	0m0.069s
sys	0m0.002s


In [ ]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input3.txt

40000000000
Executed instructions: 60
CPU times: user 6.16 ms, sys: 875 µs, total: 7.04 ms
Wall time: 106 ms


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input11.txt

39999800000
Executed instructions: 2000007
CPU times: user 5.51 ms, sys: 0 ns, total: 5.51 ms
Wall time: 204 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.006/ 2200009
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

2.727261570293576e-09
366668166.6666667


###Correctness Tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

for i in {0..10}
do
    ./max_pairwise_product_ef_cpp < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_ef.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

###Estimated maximum that can be solved in 1 second
$$(10 + 11n)*1.47*10^{-9} = 1s$$
$$ n \approx 60'000.000 $$

In [ ]:
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 60000000 <<"\n";  // n
 for (int i=1; i <60000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef_cpp

In [ ]:
!time ./max_pairwise_product_ef_cpp < input15.txt #This gives the "practical" time instead of the theoretical one (1s). The "real" constants are way bigger.

3599999940000000

real	0m26.354s
user	0m25.871s
sys	0m0.307s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("60000000\n")
        for i in range(1, 60000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
%%time
!python max_pairwise_product_ef.py < input14.txt #This gives the "practical" time instead of the theoretical one (1s). The "real" constants are way bigger.

3599999940000000
CPU times: user 328 ms, sys: 35.7 ms, total: 364 ms
Wall time: 39 s


### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    int n;
    std::cin >> n;
    int p_max = INT_MIN;
    int s_max = INT_MIN;

    for (int i = 0; i < n; i++){
        int x;
        std::cin >> x;
        if(x > p_max){
            s_max = p_max;
            p_max = x;
        }
        else if(x > s_max){
            s_max = x;
        }
    }

    std::cout << (long long) p_max * s_max << "\n";
    return 0;
}

Writing max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [ ]:
!time ./max_pairwise_product_gef_cpp < input11.txt

39999800000

real	0m0.133s
user	0m0.110s
sys	0m0.004s


20 stress tests

In [ ]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 77 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

        for (int i = 0; i < n; i++){
            int x = rand() % 1000 + 2;
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Writing stress_tests_gef_cpp.cpp


In [ ]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 52 Product: 863968
n: 59 Product: 956084
n: 53 Product: 898343
n: 22 Product: 927682
n: 2 Product: 56643
n: 32 Product: 928140
n: 48 Product: 885256
n: 26 Product: 981034
n: 40 Product: 928812
n: 23 Product: 876600
n: 55 Product: 991020
n: 27 Product: 990016
n: 11 Product: 557996
n: 70 Product: 994008
n: 53 Product: 961308
n: 58 Product: 926316
n: 39 Product: 979080
n: 3 Product: 290360
n: 23 Product: 876015
n: 26 Product: 874290


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                          //1
    ins++;
    std::cin >> n;                  //1
    ins++;
    int p_max = INT_MIN;            //1
    ins++;
    int s_max = INT_MIN;            //1
    ins++;
    ins++;
    for (int i = 0; i < n; i++){    //1 + 2n
        ins += 2;
        int x;                      //n
        ins++;
        std::cin >> x;              //n
        ins++;
        ins += 2; //both ifs
        if(x > p_max){              //[0, n]           //[0, n] means that, in the worst case, executes n times.
            s_max = p_max;          //[0, n]
            ins++;
            p_max = x;              //[0, n]
            ins++;
        }
        else if(x > s_max){         //[0, n]
            s_max = x;              //[0, n]
            ins++;
        }
    }

    ins += 3;
    std::cout << (long long) p_max * s_max << "\n"; //3 -> 8 + 4n + 5[0, n] -> 8 + 9n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Writing max_pairwise_product_gef_ins_cpp.cpp


In [ ]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!./max_pairwise_product_gef_ins_cpp < input1.txt
!./max_pairwise_product_gef_ins_cpp < input2.txt
!./max_pairwise_product_gef_ins_cpp < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52


In [ ]:
!time ./max_pairwise_product_gef_ins_cpp < input11.txt

39999800000
Executed instructions: 1600008

real	0m0.075s
user	0m0.066s
sys	0m0.001s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.0071 / 1600008
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.437477812610937e-09
225353239.4366197


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_gef_py.py
#Python3
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        x = int(i)
        if(x > p_max):
            s_max = p_max
            p_max = x
        elif x > s_max:
            s_max = x
    print(p_max * s_max)

Overwriting max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [ ]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 5.11 ms, sys: 743 µs, total: 5.85 ms
Wall time: 106 ms


In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input11.txt

39999800000
CPU times: user 5.96 ms, sys: 1.17 ms, total: 7.13 ms
Wall time: 207 ms


20 stress tests

In [ ]:
%%writefile stress_tests_gef_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        p_max = -1
        s_max = -1
        for x in range(n):
            i = random.randint(19, 133700)
            if(i > p_max):
                s_max = p_max
                p_max = i
            elif i > s_max:
                s_max = i
        print(f"n: {n} Product: {p_max * s_max}")

Writing stress_tests_gef_py.py


In [ ]:
!python stress_tests_gef_py.py

n: 6 Product: 14819845374
n: 80 Product: 17391671990
n: 74 Product: 16348413741
n: 42 Product: 17608502010
n: 65 Product: 16822345512
n: 86 Product: 17596983860
n: 80 Product: 17487831627
n: 57 Product: 14836426849
n: 58 Product: 17598723288
n: 26 Product: 16109126484
n: 43 Product: 17532080021
n: 38 Product: 14326478782
n: 73 Product: 16361114625
n: 66 Product: 16893119871
n: 49 Product: 17118182952
n: 88 Product: 17310134399
n: 9 Product: 7272889974
n: 16 Product: 17401091508
n: 73 Product: 17496624096
n: 11 Product: 11475979267


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_py.py
import random
#Python3
if __name__ == "__main__":
        ins = 0
        n = int(input())                 #1
        ins += 1
        p_max = -1                       #1
        ins += 1
        s_max = -1                       #1
        ins += 1
        ins += 1
        for i in input().split():        #1 + 3n (n reads, 2n comparisons unitl end of input)
            ins += 2
            i = int(i)                   #n
            ins += 1
            ins += 2
            if(i > p_max):               #[0, n]
                s_max = p_max            #[0, n]
                ins += 1
                p_max = i                #[0, n]
                ins += 1
            elif i > s_max:              #[0, n]
                s_max = i                #[0, n]
                ins += 1
        print(p_max * s_max)             #2
        print(f"Executed instructions: {ins}") #->6 + 4n + 5[0, n] -> 6 + 9n worst case

Writing max_pairwise_product_gef_ins_py.py


In [ ]:
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 74
90
Executed instructions: 57
40000000000
Executed instructions: 42


In [ ]:
%%time
!python max_pairwise_product_gef_ins_py.py < input11.txt

39999800000
Executed instructions: 1400004
CPU times: user 6.61 ms, sys: 0 ns, total: 6.61 ms
Wall time: 305 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.0066 / 1400004
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.714272244936443e-09
212121818.1818182


###Correctness tests

In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp

for i in {0..10}
do
    ./max_pairwise_product_gef_cpp < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

In [ ]:
%%writefile stress_test.sh
#!/bin/bash

for i in {0..10}
do
    python3 max_pairwise_product_gef_py.py < input$i.txt > ptest$i.txt
    grep -xvFf output$i.txt ptest$i.txt
done

Overwriting stress_test.sh


In [ ]:
!./stress_test.sh

###Estimated maximum that can be solved in 1 second
$$(6 + 9n)*1.47*10^{-9} = 1s$$
$$ n \approx 75'000.000 $$

In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input15.txt");

 MyFile << 75000000 <<"\n";  // n
 for (int i=1; i <75000001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest

In [ ]:
!./WriteTest

In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input15.txt #The theoretical max in 1 sec is 75M. The constants are larger in practice.

5624999925000000

real	0m33.823s
user	0m33.082s
sys	0m0.238s


In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("75000000\n")
        for i in range(1, 75000001):
            f.write(f"{i} ")

Overwriting WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input14.txt #Again, the theoretical holds. The constans are larger in practice.

5624999925000000
CPU times: user 262 ms, sys: 52.3 ms, total: 314 ms
Wall time: 38.6 s


## Perform 100 randomized automated tests in C++ and Python between the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n ln g))$.

100 stress tests sorting solution $(T(n)  = O(n ln g))$

In [ ]:
%%writefile stress_tests_sort_hundred_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 100; i++){
        int n = rand() % 69 + 2;
        std::vector<int64_t> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 10000 + 2;
        }
        std::sort(v.begin(), v.end());
        std::cout << "Test: " << i + 1 << " n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_tests_sort_hundred_cpp.cpp


In [ ]:
!g++ stress_tests_sort_hundred_cpp.cpp -o hundred_cpp
!./hundred_cpp

Test: 1 n: 36 Product: 89941896
Test: 2 n: 57 Product: 98922772
Test: 3 n: 60 Product: 94099610
Test: 4 n: 31 Product: 97061823
Test: 5 n: 13 Product: 83663118
Test: 6 n: 51 Product: 97087172
Test: 7 n: 12 Product: 88055612
Test: 8 n: 26 Product: 90714270
Test: 9 n: 70 Product: 97207578
Test: 10 n: 14 Product: 82659976
Test: 11 n: 61 Product: 98703600
Test: 12 n: 52 Product: 77473704
Test: 13 n: 10 Product: 73179975
Test: 14 n: 32 Product: 92630928
Test: 15 n: 21 Product: 91131372
Test: 16 n: 24 Product: 79372320
Test: 17 n: 2 Product: 6315562
Test: 18 n: 54 Product: 92974560
Test: 19 n: 8 Product: 97712784
Test: 20 n: 23 Product: 98495568
Test: 21 n: 36 Product: 75369105
Test: 22 n: 47 Product: 95000409
Test: 23 n: 10 Product: 53909000
Test: 24 n: 46 Product: 97859750
Test: 25 n: 30 Product: 85945648
Test: 26 n: 63 Product: 99349934
Test: 27 n: 69 Product: 90139400
Test: 28 n: 67 Product: 95317922
Test: 29 n: 14 Product: 92507982
Test: 30 n: 27 Product: 70234203
Test: 31 n: 26 Product

In [ ]:
%%writefile stress_tests_sort_hundred_py.py
#Python3
import random
if __name__ == "__main__":
    for i in range(100):
        n = random.randint(4, 186)
        l = [ random.randint(27, 56734) for i in range(n) ]
        l.sort()
        print(f"Test: {i + 1} n: {n} Product: {l[-1] * l[-2]}")

Writing stress_tests_sort_hundred_py.py


In [ ]:
!python stress_tests_sort_hundred_py.py

Test: 1 n: 12 Product: 2667432826
Test: 2 n: 161 Product: 3205257136
Test: 3 n: 152 Product: 3210694125
Test: 4 n: 56 Product: 3166559928
Test: 5 n: 134 Product: 3158963609
Test: 6 n: 137 Product: 3165037445
Test: 7 n: 55 Product: 2902371984
Test: 8 n: 139 Product: 3184231897
Test: 9 n: 65 Product: 3020854085
Test: 10 n: 83 Product: 3130961098
Test: 11 n: 134 Product: 3193606128
Test: 12 n: 78 Product: 3183426459
Test: 13 n: 97 Product: 3052044684
Test: 14 n: 90 Product: 3143077812
Test: 15 n: 131 Product: 3078388670
Test: 16 n: 10 Product: 1965848421
Test: 17 n: 101 Product: 3157183840
Test: 18 n: 104 Product: 3180223060
Test: 19 n: 152 Product: 3177971500
Test: 20 n: 43 Product: 3132604036
Test: 21 n: 60 Product: 3072829150
Test: 22 n: 77 Product: 3142606636
Test: 23 n: 40 Product: 2781426944
Test: 24 n: 114 Product: 3178164420
Test: 25 n: 13 Product: 3126437688
Test: 26 n: 54 Product: 2900777875
Test: 27 n: 110 Product: 3185245908
Test: 28 n: 37 Product: 2934317400
Test: 29 n: 40 Pr

100 stress tests super efficient solution $(T(n) < 10 n = O(n))$

In [ ]:
%%writefile stress_tests_gef_hundred_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    for(int i = 0; i < 100; i++){
        int n = rand() % 93 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

        for (int i = 0; i < n; i++){
            int64_t x = rand() % 16004 + 2;
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "Test: " << i + 1 << " n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Writing stress_tests_gef_hundred_cpp.cpp


In [ ]:
!g++ stress_tests_gef_hundred_cpp.cpp -o hundred_gef_cpp
!./hundred_gef_cpp

Test: 1 n: 12 Product: 236744356
Test: 2 n: 39 Product: 204588032
Test: 3 n: 22 Product: 241548723
Test: 4 n: 10 Product: 170150412
Test: 5 n: 73 Product: 248927946
Test: 6 n: 78 Product: 248229138
Test: 7 n: 79 Product: 246900555
Test: 8 n: 11 Product: 171471131
Test: 9 n: 67 Product: 248089160
Test: 10 n: 5 Product: 155684700
Test: 11 n: 92 Product: 254322726
Test: 12 n: 37 Product: 244301265
Test: 13 n: 62 Product: 221214448
Test: 14 n: 6 Product: 203766624
Test: 15 n: 30 Product: 218059776
Test: 16 n: 19 Product: 207585968
Test: 17 n: 59 Product: 247228600
Test: 18 n: 78 Product: 252868563
Test: 19 n: 78 Product: 251076116
Test: 20 n: 60 Product: 220715592
Test: 21 n: 81 Product: 246637421
Test: 22 n: 56 Product: 247481183
Test: 23 n: 64 Product: 236522858
Test: 24 n: 45 Product: 220676568
Test: 25 n: 28 Product: 215968320
Test: 26 n: 85 Product: 246050547
Test: 27 n: 78 Product: 253365246
Test: 28 n: 88 Product: 244093696
Test: 29 n: 89 Product: 254450296
Test: 30 n: 90 Product: 2

In [ ]:
%%writefile stress_tests_gef_hundred_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(100):
        n = random.randint(5, 456)
        p_max = -1
        s_max = -1
        for j in range(n):
            x = random.randint(19, 13376)
            if(x > p_max):
                s_max = p_max
                p_max = x
            elif x > s_max:
                s_max = x
        print(f"Test: {i + 1} n: {n} Product: {p_max * s_max}")

Writing stress_tests_gef_hundred_py.py


In [ ]:
!python stress_tests_gef_hundred_py.py

Test: 1 n: 187 Product: 174617754
Test: 2 n: 268 Product: 178556394
Test: 3 n: 299 Product: 177421311
Test: 4 n: 285 Product: 176810145
Test: 5 n: 62 Product: 175159141
Test: 6 n: 311 Product: 178810375
Test: 7 n: 387 Product: 177502313
Test: 8 n: 312 Product: 177622256
Test: 9 n: 436 Product: 177528976
Test: 10 n: 227 Product: 176592192
Test: 11 n: 153 Product: 178502888
Test: 12 n: 411 Product: 176782460
Test: 13 n: 195 Product: 174290955
Test: 14 n: 228 Product: 177794880
Test: 15 n: 380 Product: 177036058
Test: 16 n: 113 Product: 172114800
Test: 17 n: 167 Product: 173923800
Test: 18 n: 278 Product: 178596415
Test: 19 n: 28 Product: 174055248
Test: 20 n: 253 Product: 177675108
Test: 21 n: 386 Product: 178262546
Test: 22 n: 363 Product: 176690176
Test: 23 n: 388 Product: 176514660
Test: 24 n: 331 Product: 177715512
Test: 25 n: 250 Product: 177741740
Test: 26 n: 386 Product: 178034688
Test: 27 n: 85 Product: 157557782
Test: 28 n: 141 Product: 176902958
Test: 29 n: 64 Product: 17482080

## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

#Random tests input

In [ ]:
%%writefile rgt.cpp
#include <bits/stdc++.h>

int main(int argc, char* argv[]){
    std::random_device rd;
    std::mt19937_64 rng(rd());
    std::uniform_int_distribution<> ls_dst(2, 100);

    std::ofstream input("hundred_rtests.txt");
    for(int i = 0; i < 100; i++){
        int64_t n = ls_dst(rng);
        input << n << ' ';
        while(n--){
            std::uniform_int_distribution<> dst(1337, 400000);
            input << dst(rng) << ' ';
        }
        input << '\n';
    }
    return 0;
}

Overwriting rgt.cpp


In [ ]:
!g++ rgt.cpp -o rgt
!./rgt

#C++. O(nlogn) - O(n)

#O(nlogn)

In [ ]:
%%writefile hundred_rtests_sort_cpp.cpp
#include <bits/stdc++.h>

int main(int argc, char* argv[]){
    std::ifstream input("hundred_rtests.txt");
    std::ofstream output("hundred_rtest_sort.txt");
    std::string l;
    int64_t n;
    clock_t s = std::clock();
    while(std::getline(input, l)){
        std::istringstream iss(l);
        iss >> n;
        int64_t x;
        std::vector<int64_t> v;
        while(iss >> x){ v.push_back(x); }
        std::sort(v.begin(), v.end());
        int64_t ans = v[n - 1] * v[n - 2];
        output << ans << '\n';
    }
    clock_t e = std::clock();
    std::cout << "Time elapsed: " << (double)(e - s) / CLOCKS_PER_SEC << '\n';
    return 0;
}

Overwriting hundred_rtests_sort_cpp.cpp


In [ ]:
!g++ hundred_rtests_sort_cpp.cpp -o hundred_rtests_sort_cpp
!./hundred_rtests_sort_cpp

Time elapsed: 0.002131


#O(n)

In [ ]:
%%writefile hundred_rtests_gef_cpp.cpp
#include <bits/stdc++.h>

int main(int argc, char* argv[]){
    std::ifstream input("hundred_rtests.txt");
    std::ofstream output("hundred_rtest_gef.txt");
    std::string l;
    int64_t n;
    clock_t s = std::clock();
    while(std::getline(input, l)){
        std::istringstream iss(l);
        iss >> n;
        int64_t x;
        int64_t p_max = INT_MIN;
        int64_t s_max = INT_MIN;
        while(iss >> x){
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }
        int64_t ans = p_max * s_max;
        output << ans << '\n';
    }
    clock_t e = std::clock();
    std::cout << "Time elapsed: " << (double)(e - s) / CLOCKS_PER_SEC << '\n';
    return 0;
}

Overwriting hundred_rtests_gef_cpp.cpp


In [ ]:
!g++ hundred_rtests_gef_cpp.cpp -o hundred_rtests_gef_cpp
!./hundred_rtests_gef_cpp

Time elapsed: 0.000823


Checking for consistency in each algorithm output

In [ ]:
%%writefile cpp_hundred.sh
#!/bin/bash
grep -xvFf hundred_rtest_sort.txt hundred_rtest_gef.txt

Overwriting cpp_hundred.sh


In [ ]:
#if no output from grep, same values in each txt file i.e. same results in each test case
!chmod 744 cpp_hundred.sh
!./cpp_hundred.sh

#Python. O(nlogn) - O(n)

#O(nlogn)

In [ ]:
%%writefile hundred_rtests_sort_py.py
if __name__ == "__main__":
    with open("hundred_rtests_sort_py.txt", 'w') as output:
        with open("hundred_rtests.txt", 'r') as inp:
            for l in inp:
                v = []
                v = [ int(i) for i in l.split() ]
                v.pop(0)
                v.sort()
                ans = v[-1] * v[-2]
                output.write(f"{ans}\n")

Overwriting hundred_rtests_sort_py.py


In [ ]:
%%time
#in ms
!python hundred_rtests_sort_py.py

CPU times: user 6.59 ms, sys: 0 ns, total: 6.59 ms
Wall time: 109 ms


#O(n)

In [ ]:
%%writefile hundred_rtests_gef_py.py
if __name__ == "__main__":
    with open("hundred_rtests_gef_py.txt", 'w') as output:
        with open("hundred_rtests.txt", 'r') as inp:
            for l in inp:
                l = list(map(int, l.split()))
                n = l[0]
                p_max = -1
                s_max = -1
                for x in l[1:]:
                    if(x > p_max):
                        s_max = p_max
                        p_max = x
                    elif x > s_max:
                        s_max = x
                ans = p_max * s_max
                output.write(f"{ans}\n")

Overwriting hundred_rtests_gef_py.py


In [ ]:
%%time
#in ms
!python hundred_rtests_gef_py.py

CPU times: user 6.22 ms, sys: 0 ns, total: 6.22 ms
Wall time: 106 ms


Checking for consistency in each algorithm output

In [ ]:
#if no output from grep, same values in each txt file i.e. same results in each test case
%%writefile py_hundred.sh
#!/bin/bash
grep -xvFf hundred_rtests_sort_py.txt hundred_rtests_gef_py.txt

Overwriting py_hundred.sh


In [ ]:
!chmod 744 py_hundred.sh
!./py_hundred.sh

#Cross checking
Python O(nlogn) - C++ O(nlogn)<br />
Python O(nlogn) - C++ O(n)<br />
Python O(n) - C++ O(nlogn)<br />
Python O(n) - C++ O(n)

In [ ]:
#if no output from grep, same values in each txt file i.e. same results in each test case
%%writefile hundred_cross_checking.sh
#!/bin/bash
grep -xvFf hundred_rtests_sort_py.txt hundred_rtest_sort.txt
grep -xvFf hundred_rtests_sort_py.txt hundred_rtest_gef.txt
grep -xvFf hundred_rtests_gef_py.txt hundred_rtest_sort.txt
grep -xvFf hundred_rtests_gef_py.txt hundred_rtest_gef.txt

Overwriting hundred_cross_checking.sh


In [ ]:
!chmod 744 hundred_cross_checking.sh
!./hundred_cross_checking.sh